File: scripts/oop-practice-sources/lua/student.lua
local Student = {PASSING_SCORE = 60}

function Student:new(name)
    local obj = setmetatable({name = name, grades = {}}, {__index = self})
    return obj
end

function Student:add_grade(grade)
    table.insert(self.grades, grade)
    print("Оценка " .. grade .. " добавлена для " .. self.name)
end

function Student:average_score()
    if #self.grades == 0 then
        return 0
    end
    local sum = 0
    for _, g in ipairs(self.grades) do
        sum = sum + g
    end
    return sum / #self.grades
end

function Student:is_passing()
    return self:average_score() >= self.PASSING_SCORE
end

function Student:info()
    local avg = self:average_score()
    print("Студент: " .. self.name)
    print("Оценки: " .. table.concat(self.grades, ", "))
    print(string.format("Средний балл: %.1f", avg))
    if self:is_passing() then
        print("Зачёт получен")
    else
        print("Зачёт не получен")
    end
end

local student = Student:new("Анна")
student:add_grade(70)
student:add_grade(85)
student:add_grade(55)
student:info()

File: scripts/oop-practice-sources/rust/student.rs
struct Student {
    name: String,
    grades: Vec<i32>,
}

impl Student {
    const PASSING_SCORE: i32 = 60;

    fn new(name: &str) -> Self {
        Self {
            name: name.to_string(),
            grades: Vec::new(),
        }
    }

    fn add_grade(&mut self, grade: i32) {
        self.grades.push(grade);
        println!("Оценка {} добавлена для {}", grade, self.name);
    }

    fn average_score(&self) -> f64 {
        if self.grades.is_empty() {
            return 0.0;
        }
        let sum: i32 = self.grades.iter().sum();
        sum as f64 / self.grades.len() as f64
    }

    fn is_passing(&self) -> bool {
        self.average_score() >= Self::PASSING_SCORE as f64
    }

    fn info(&self) {
        println!("Студент: {}", self.name);
        println!("Оценки: {:?}", self.grades);
        println!("Средний балл: {:.1}", self.average_score());
        if self.is_passing() {
            println!("Зачёт получен");
        } else {
            println!("Зачёт не получен");
        }
    }
}

fn main() {
    let mut student = Student::new("Анна");
    student.add_grade(70);
    student.add_grade(85);
    student.add_grade(55);
    student.info();
}
